cobc/dump_tree.c

```diff
@@ -0,0 +1,568 @@
+		"| FLAGS of string list\n"
+		"let format_date = \"GNUCOBOL-2023-07-15\"\n"
+		"\n"
+		"let tree =\n"
+		;
+
+	fmt.format_trailer = "\n"
+		"let () =\n"
+		"  let oc = open_out_bin \"tree.cbb\" in\n"
+		"  output_value oc ( format_date : string );\n"
+		"  output_value oc ( tree : t);\n"
+		"  close_out oc\n"
+		"\n";
+	fmt.format_begin_field = "\"";
+	fmt.format_end_field = "\", ";
+
+	fmt.format_begin_record = "RECORD [";
+	fmt.format_delim_record = ";" ;
+	fmt.format_last_delim_record = ";";
+	fmt.format_end_record = "]";
+
+	fmt.format_begin_list = "LIST [";
+	fmt.format_delim_list =  ";";
+	fmt.format_last_delim_list =  ";";
+	fmt.format_end_list = "]";
+
+	fmt.format_null =   "NULL";
+	fmt.format_begin_constr = "CONSTR (\"";
+	fmt.format_delim_constr = "\", ";
+	fmt.format_end_constr = ")";
+	fmt.format_begin_pointer = "POINTER ";
+	fmt.format_end_pointer = "L";
+	fmt.format_begin_char = "CHAR '";
+	fmt.format_end_char = "'";
+	fmt.format_begin_string = "STRING \"";
+	fmt.format_end_string = "\"";
+	fmt.format_begin_int = "INT( ";
+	fmt.format_end_int = ")";
+
+	fmt.format_begin_flags = "FLAGS [";
+	fmt.format_begin_flag = "\"";
+	fmt.format_end_flag = "\";";
+	fmt.format_end_flags = "]";
+}
+
+static void set_json_format (void)
+{
+	/* Default is to use JSON */
+	/* Standard JSON format. Tests validated by jsonlint-php. */
+
+	fmt.format_kind = FORMAT_JSON;
+	fmt.format_header = "";
+	fmt.format_trailer = "";
+
+	fmt.format_begin_field = "\"";
+	fmt.format_end_field = "\": ";
+
+	fmt.format_begin_record = "{";
+	fmt.format_delim_record = ",";
+	fmt.format_last_delim_record = "";
+	fmt.format_end_record = "}";
+
+	fmt.format_begin_list = "[";
+	fmt.format_delim_list = ",";
+	fmt.format_last_delim_list = "";
+	fmt.format_end_list = "]";
+
+	fmt.format_begin_flags = "[";
+	fmt.format_begin_flag = "\"";
+	fmt.format_end_flag = "\",";
+	fmt.format_end_flags = " NULL]";
+
+	fmt.format_null = "null";
+	fmt.format_begin_constr = "{ \"type_\": \"";
+	fmt.format_delim_constr = "\", \"value_\": ";
+	fmt.format_end_constr = " }";
+	fmt.format_begin_pointer = "\"";
+	fmt.format_end_pointer = "\"";
+	fmt.format_begin_char =  "\"";
+	fmt.format_end_char = "\"";
+	fmt.format_begin_string =  "\"";
+	fmt.format_end_string = "\"";
+	fmt.format_begin_int = "";
+	fmt.format_end_int = "";
+}
+
+static
+void set_format_by_file_ext (const char* filename)
+{
+	const char *arg = filename + strlen(filename);
+
+	while ( arg > filename && *arg != '.' ) arg--;
+
+	if (!strcasecmp(arg, ".ml")){
+
+		set_ocaml_format ();
+
+		cb_dump_tree_with_common = 1;
+		cb_dump_tree_with_loc = 1;
+		cb_dump_tree_with_type = 1;
+		cb_dump_tree_with_pointer = 1;
+		cb_dump_tree_with_indent = 1;
+		cb_dump_tree_with_newlines = 1;
+
+		return;
+	}
+
+	set_json_format ();
+}
+
+static
+void set_flags (const char* flags)
+{
+	int sign = 1;
+	int i;
+	int len = strlen(flags);
+
+	for(i=0; i< len; i++){
+		switch(flags[i]){
+		case '+': sign = 1; break;
+		case '-': sign = 0; break;
```

Comment:
these flags duplicate the filename option and should therefore be dropped here and in the flag.def help

Reply:
The goal is to be able to enable/disable flags in the same definition, for example `+z-i` (add zero-value fields and disable indent)